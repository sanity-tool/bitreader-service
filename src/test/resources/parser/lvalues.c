static void localAuto() {
    int a = 1;
}

static void localStatic() {
    static int a = 1;
    a = 2;
}

static int ARRAY[10];

static void arrayElementAccess() {
    ARRAY[0] = 1;
    ARRAY[1] = 2;
}

static void arrayDeref() {
    *ARRAY = 1;
}

static int *PTR;

static void ptrElementAccess() {
    PTR[0] = 1;
    PTR[1] = 2;
}

static void ptrDeref() {
    *PTR = 1;
}

struct {
    int field;
} STRUCT, *PTR_STRUCT;

static void structFieldAccess() {
    STRUCT.field = 1;
}

static void structPtrFieldAccess() {
    (*PTR_STRUCT).field = 1;
    PTR_STRUCT->field = 2;
}