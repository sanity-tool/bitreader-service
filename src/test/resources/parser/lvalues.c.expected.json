{
  "functions" : [ {
    "ref" : {
      "kind" : "Global",
      "index" : 0
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMAlloca",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        } ]
      }, {
        "kind" : "LLVMCall",
        "operands" : [ {
          "kind" : "LLVMMetadataAsValueValueKind",
          "typeId" : 4
        }, {
          "kind" : "LLVMMetadataAsValueValueKind",
          "typeId" : 4
        }, {
          "kind" : "LLVMMetadataAsValueValueKind",
          "typeId" : 4
        }, {
          "kind" : "LLVMFunctionValueKind",
          "name" : "llvm.dbg.declare",
          "typeId" : 5
        } ]
      }, {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  }, {
    "ref" : {
      "kind" : "Global",
      "index" : 1
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMGlobalVariableValueKind",
          "name" : "localStatic.a",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  }, {
    "ref" : {
      "kind" : "Global",
      "index" : 2
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMConstantExprValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMConstantExprValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  }, {
    "ref" : {
      "kind" : "Global",
      "index" : 3
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMConstantExprValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  }, {
    "ref" : {
      "kind" : "Global",
      "index" : 4
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMLoad",
        "operands" : [ {
          "kind" : "LLVMGlobalVariableValueKind",
          "name" : "PTR",
          "typeId" : 8
        } ]
      }, {
        "kind" : "LLVMGetElementPtr",
        "operands" : [ {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        }, {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 9
        } ]
      }, {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMLoad",
        "operands" : [ {
          "kind" : "LLVMGlobalVariableValueKind",
          "name" : "PTR",
          "typeId" : 8
        } ]
      }, {
        "kind" : "LLVMGetElementPtr",
        "operands" : [ {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        }, {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 9
        } ]
      }, {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  }, {
    "ref" : {
      "kind" : "Global",
      "index" : 5
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMLoad",
        "operands" : [ {
          "kind" : "LLVMGlobalVariableValueKind",
          "name" : "PTR",
          "typeId" : 8
        } ]
      }, {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  }, {
    "ref" : {
      "kind" : "Global",
      "index" : 6
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMConstantExprValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  }, {
    "ref" : {
      "kind" : "Global",
      "index" : 7
    },
    "blocks" : [ {
      "instructions" : [ {
        "kind" : "LLVMLoad",
        "operands" : [ {
          "kind" : "LLVMGlobalVariableValueKind",
          "name" : "PTR_STRUCT",
          "typeId" : 10
        } ]
      }, {
        "kind" : "LLVMGetElementPtr",
        "operands" : [ {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 11
        }, {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        } ]
      }, {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMLoad",
        "operands" : [ {
          "kind" : "LLVMGlobalVariableValueKind",
          "name" : "PTR_STRUCT",
          "typeId" : 10
        } ]
      }, {
        "kind" : "LLVMGetElementPtr",
        "operands" : [ {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 11
        }, {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        } ]
      }, {
        "kind" : "LLVMStore",
        "operands" : [ {
          "kind" : "LLVMConstantIntValueKind",
          "typeId" : 3
        }, {
          "kind" : "LLVMInstructionValueKind",
          "typeId" : 7
        } ]
      }, {
        "kind" : "LLVMRet"
      } ]
    } ],
    "entryBlockIndex" : 0
  } ],
  "types" : [ {
    "kind" : "LLVMPointerTypeKind",
    "types" : [ 1 ]
  }, {
    "kind" : "LLVMFunctionTypeKind",
    "types" : [ 2 ]
  }, {
    "kind" : "LLVMVoidTypeKind"
  }, {
    "kind" : "LLVMIntegerTypeKind",
    "len" : 32
  }, {
    "kind" : "LLVMMetadataTypeKind"
  }, {
    "kind" : "LLVMPointerTypeKind",
    "types" : [ 6 ]
  }, {
    "kind" : "LLVMFunctionTypeKind",
    "types" : [ 2, 4, 4, 4 ]
  }, {
    "kind" : "LLVMPointerTypeKind",
    "types" : [ 3 ]
  }, {
    "kind" : "LLVMPointerTypeKind",
    "types" : [ 7 ]
  }, {
    "kind" : "LLVMIntegerTypeKind",
    "len" : 64
  }, {
    "kind" : "LLVMPointerTypeKind",
    "types" : [ 11 ]
  }, {
    "kind" : "LLVMPointerTypeKind",
    "types" : [ 12 ]
  }, {
    "kind" : "LLVMStructTypeKind",
    "name" : "struct.anon",
    "types" : [ 3 ]
  } ]
}